### PR TITLE
normalize adt fields during structural match checking

### LINKED ISSUE #72896: ICE when structurally matching a struct containing an associated type on latest Nightly.
One of the tests in my enumset crate has started throwing an ICE on the latest nightly.

The problem seems to be that I use an associated type as a field in a struct I'm structurally matching, and this has been broken on the latest nightly.

### Code

[Playground Link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=05bb91b14fad31c31f66b4409abd43a7)

```Rust
trait EnumSetType: Copy + Eq {
    type Repr;
}

#[derive(Copy, Clone, PartialEq, Eq)]
enum Enum8 { }
impl EnumSetType for Enum8 {
    type Repr = u8;
}

#[derive(Copy, Clone, PartialEq, Eq)]
struct EnumSet<T: EnumSetType> {
    __enumset_underlying: T::Repr,
}

const CONST_SET: EnumSet<Enum8> = EnumSet { __enumset_underlying: 3 };

#[test]
fn match_const_test() {
    match CONST_SET {
        CONST_SET => { /* ok */ }
        _ => panic!("match fell through?"),
    }
}
```

### Error output

```
   Compiling test_crate v0.1.0 (/home/[...]/enumset_test)
error: internal compiler error: src/librustc_mir_build/hair/pattern/const_to_pat.rs:137: use of a constant whose type is a projection inside a pattern

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:907:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.45.0-nightly (5fd2f06e9 2020-05-31) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

error: aborting due to previous error

error: could not compile `test_crate`.

To learn more, run the command again with --verbose.
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
   Compiling test_crate v0.1.0 (/home/[...]/enumset_test)
error: internal compiler error: src/librustc_mir_build/hair/pattern/const_to_pat.rs:137: use of a constant whose type is a projection inside a pattern

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:907:9
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1076
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1537
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:490
  12: std::panicking::begin_panic
  13: rustc_errors::HandlerInner::bug
  14: rustc_errors::Handler::bug
  15: rustc_middle::util::bug::opt_span_bug_fmt::{{closure}}
  16: rustc_middle::ty::context::tls::with_opt::{{closure}}
  17: rustc_middle::ty::context::tls::with_opt
  18: rustc_middle::util::bug::opt_span_bug_fmt
  19: rustc_middle::util::bug::bug_fmt
  20: rustc_mir_build::hair::pattern::const_to_pat::ConstToPat::to_pat
  21: rustc_middle::ty::context::GlobalCtxt::enter_local
  22: rustc_mir_build::hair::pattern::PatCtxt::lower_path
  23: rustc_mir_build::hair::pattern::PatCtxt::lower_pattern
  24: rustc_mir_build::hair::pattern::check_match::MatchVisitor::lower_pattern
  25: <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
  26: <rustc_mir_build::hair::pattern::check_match::MatchVisitor as rustc_hir::intravisit::Visitor>::visit_expr
  27: <rustc_mir_build::hair::pattern::check_match::MatchVisitor as rustc_hir::intravisit::Visitor>::visit_expr
  28: rustc_mir_build::hair::pattern::check_match::check_match
  29: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::check_match>::compute
  30: rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps
  31: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  32: rustc_data_structures::stack::ensure_sufficient_stack
  33: rustc_query_system::query::plumbing::get_query_impl
  34: rustc_query_system::query::plumbing::ensure_query_impl
  35: rustc_session::utils::<impl rustc_session::session::Session>::time
  36: rustc_session::utils::<impl rustc_session::session::Session>::time
  37: rustc_interface::passes::analysis
  38: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::analysis>::compute
  39: rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps
  40: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  41: rustc_query_system::query::plumbing::get_query_impl
  42: rustc_middle::ty::context::tls::enter_global
  43: rustc_interface::interface::run_compiler_in_existing_thread_pool
  44: rustc_ast::attr::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.45.0-nightly (5fd2f06e9 2020-05-31) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [check_match] processing `match_const_test`
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to previous error

error: could not compile `test_crate`.

To learn more, run the command again with --verbose.
```
</p>
</details>

### PR DESCRIPTION
**Pull request by lcnr**
fixes #72896 

currently only fixes the issue itself and compiles stage 1 libs.
I believe we have to use something else to normalize the adt fields here,
as I expect some partially resolved adts to cause problems :thinking: 

stage 1 libs and the test itself pass, not sure about the rest...
Will spend some more time looking into it tomorrow.

r? @pnkfelix cc @eddyb

### CODE REVIEW

**Comment by lcnr** [comment]
Apparently this works even if I believe it shouldn't...
Should now at least be ready for review I guess :shrug:

File: src/librustc_trait_selection/traits/structural_match.rs [hunk 1]
```diff
@@ -251,7 +251,10 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
         // fields of ADT.
         let tcx = self.tcx();
         for field_ty in adt_def.all_fields().map(|field| field.ty(tcx, substs)) {
-            if field_ty.visit_with(self) {
+            let ty = self.tcx().normalize_erasing_regions(ty::ParamEnv::empty(), field_ty);
```
Comment (eddyb) [inline src/librustc_trait_selection/traits/structural_match.rs:254]: `ty::Param::empty()` is incorrect - frankly I wonder if we should remove it.
You should have a `ParamEnv` around.

(*couple minutes later*)

Hang on, this entire file is missing real `ParamEnv`s? That's liable to cause ICEs if generic parameters end up in these types.

Comment (lcnr) [reply src/librustc_trait_selection/traits/structural_match.rs:254]: Does this block this PR? Afaik it's currently not possible on stable to hit this case.

Comment (lcnr) [reply src/librustc_trait_selection/traits/structural_match.rs:254]: I may go ahead and fix this afterwards, but my current priority is mainly to fix #72896 
Which probably means that we have to backport this to beta.

Another option would be to revert the structural match check for projections :thinking:

**Comment by pnkfelix** [comment]
@bors r+

Lets land this band-aid now, even with the `ty::ParamEnv::empty()` there. We can work on fixing the file over all as a longer-term thing.

File: src/librustc_trait_selection/traits/structural_match.rs (hunk 1 above)
Comment (pnkfelix) [reply src/librustc_trait_selection/traits/structural_match.rs:254]: Its probably my fault that we don't have `ParamEnv`s threaded through here. This whole thing of switching from attributes to traits has been a real rabbit hole.

Still, I think landing this is better than reverting the structural match check for projections.

**Comment by pnkfelix** [comment]
Discussed in [todays T-compiler meeting](https://zulip-archive.rust-lang.org/238009tcompilermeetings/72518weeklymeeting2020061154818.html#200549162). Approved for beta-backport.

**Pull request merged**